libqtile/backend/wayland/qw/cairo-buffer.c

```diff
@@ -51,3 +51,66 @@ struct wlr_buffer *cairo_buffer_create(int width, int height, size_t stride, voi
     cairo_buffer->stride = stride;
     return &cairo_buffer->base;
 }
+
+// Creates a wlr_buffer from a subregion of a cairo image surface
+struct wlr_buffer *cairo_buffer_from_surface_region(
+    cairo_surface_t *surface,
+    int x, int y, int width, int height
+) {
+    if (cairo_surface_get_type(surface) != CAIRO_SURFACE_TYPE_IMAGE) {
+        wlr_log(WLR_ERROR, "Cairo surface is not an image surface\n");
+        return NULL;
+    }
+
+    cairo_surface_flush(surface);
+
+    unsigned char *data = cairo_image_surface_get_data(surface);
+    int surface_width = cairo_image_surface_get_width(surface);
+    int surface_height = cairo_image_surface_get_height(surface);
+    int stride = cairo_image_surface_get_stride(surface);
+
+    if (x < 0 || y < 0 || x + width > surface_width || y + height > surface_height) {
```

Comment:
Shouldn't the bound check use `>=` if it starts at 0?
```suggestion
    if (x < 0 || y < 0 || x + width >= surface_width || y + height >= surface_height) {
```

Reply:
No. I don't think so. x+width should equal surface_width for a region touching the right hand side.

If I make the change, it does trigger the error.

Reply:
so x == width is a valid case?